docs/ilab-model-engine.md

```diff
@@ -0,0 +1,160 @@
+
+```console
+ilab
+|
+|_______download__|
+|_______convert___|-----model
+|_______train_____|
+```
+
+This structure works when commands are ALL linked to the same `verb`. What happens when a group like `download` gets a command that the others don't? Then it looks more like this:
+
+```console
+ilab
+|
+|_______download
+|       |____model
+|       |____taxonomy
+|_______convert
+|       |_____model
+|_______train
+|       |_____model
+```
+
+This is opposed to the structure in this EP, which might have a duplicate here and there for something like `download`, but in this structure duplicate commands are the norm and will result in a larger and clunkier codebase that is confusing to read.
+
+This first structure looks nice, and `ilab download model` sounds nice. However in terms of implementation, this makes little sense. In [click](https://click.palletsprojects.com/en/8.1.x/), the CLI library we use, `ilab` is a "group" that commands are currently "grouped under". In this Model, `ilab` would be a group, `download`, `convert` etc would be a sub-group, and `model` would be a command under ALL of these groups. This would require an implementation of the `model` function in different packages all linking to a different group parent. From the perspective of user contributions and general code readability doesn't make much sense. Part of this design is to encourage user contributions by making the structure of the codebase logical. While these commands make the commands sound better, they make them harder to group, and understand.
+
+### General workflow as compared to alternatives
+
+currently one has to:
+
+1. `ilab init`
+2. `ilab download`
+3. `ilab serve`
+4. `ilab generate`
+5. `ilab train`
+6. `ilab convert`
+7. `ilab serve --model XXX`
+8. `ilab chat`
+
+how to use `ilab` with this new structure:
+
+1. `ilab config init`
+2. `ilab model download`
+3. `ilab data generate`
+4. `ilab model train --convert`
+5. `ilab model serve -i` (or followed by `ilab model chat`)
+
+So, not only do we remove one of the steps (with `-i`) for most users, but it is actually clear what is happening. For example, one may ask: "What does `ilab init` do?" while `ilab config init` clearly initializes the CLI's config. A more clear example is `ilab generate`. Are we generating a model? A config? No, we are generating *data*: `ilab data generate`.
+
+## Changes to Existing flow
+
+The current `ilab` commands will still work. Users will be able to type commands like `ilab model train` or `ilab train` for the foreseeable future to ensure feature parity. Eventually, this alias should be removed, and only the sub-commands should probably exist.
+
+## Known Issues
+
+### click
+
+click doesn't like the setup we currently have in `ilab`. So, adding sub-parent commands won't be as easy as creating a new `click.group`. We will need to make different libraries, each of which is a click group most likely.
```

Comment:
If click is bad at this, is argparse better?

Also the "different libraries" comment is not clear to me and makes me concerned about how restructuring for click might be different than restructuring for future API or for general good structure.

Reply:
@cdoern I second that using groups might not be straightforward and the current `lab.py` might require a bit of remodeling but I think the statement "adding sub-parent commands won't be as easy as creating a new `click.group`" is a bit misleading. Implementing something like `ilab foo bar --baz` is possible :). I'm not sure why we need to make different libraries. Thanks!